engine/src/main/java/com/agiletec/aps/system/services/category/cache/CategoryManagerCacheWrapper.java

```diff
@@ -0,0 +1,151 @@
+
+import com.agiletec.aps.system.common.AbstractCacheWrapper;
+import com.agiletec.aps.system.exception.ApsSystemException;
+import com.agiletec.aps.system.services.category.Category;
+import com.agiletec.aps.system.services.category.ICategoryDAO;
+import com.agiletec.aps.system.services.lang.ILangManager;
+import com.agiletec.aps.system.services.lang.Lang;
+import com.agiletec.aps.util.ApsProperties;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.cache.Cache;
+
+/**
+ * @author E.Santoboni
+ */
+public class CategoryManagerCacheWrapper extends AbstractCacheWrapper implements ICategoryManagerCacheWrapper {
+
+	private static final Logger _logger = LoggerFactory.getLogger(CategoryManagerCacheWrapper.class);
+
+	@Override
+	public void initCache(ICategoryDAO categoryDAO, ILangManager langManager) throws ApsSystemException {
+		List<Category> categories = null;
+		try {
+			categories = categoryDAO.loadCategories(langManager);
+			if (categories.isEmpty()) {
+				Category root = this.createRoot(langManager);
+				categoryDAO.addCategory(root);
+				categories.add(root);
+			}
+			this.initCache(categories);
+		} catch (Throwable t) {
+			_logger.error("Error loading the category tree", t);
+			throw new ApsSystemException("Error loading the category tree.", t);
+		}
+	}
+
+	protected Category createRoot(ILangManager langManager) {
+		Category root = new Category();
+		root.setCode("home");
+		root.setParentCode("home");
+		List<Lang> langs = langManager.getLangs();
+		ApsProperties titles = new ApsProperties();
+		for (int i = 0; i < langs.size(); i++) {
+			Lang lang = (Lang) langs.get(i);
+			titles.setProperty(lang.getCode(), "Home");
+		}
+		root.setTitles(titles);
+		return root;
+	}
+
+	private void initCache(List<Category> categories) throws ApsSystemException {
+		Cache cache = this.getCache();
+		this.releaseCachedObjects(cache);
+		Category root = null;
+		Map<String, Category> categoryMap = new HashMap<String, Category>();
+		for (int i = 0; i < categories.size(); i++) {
+			Category cat = (Category) categories.get(i);
+			categoryMap.put(cat.getCode(), cat);
+			if (cat.getCode().equals(cat.getParentCode())) {
+				root = cat;
+			}
+		}
+		for (int i = 0; i < categories.size(); i++) {
+			Category cat = categories.get(i);
+			Category parent = categoryMap.get(cat.getParentCode());
+			if (cat != root) {
+				parent.addChildCode(cat.getCode());
+			}
+			cat.setParent(parent);
+		}
+		if (root == null) {
+			throw new ApsSystemException("Error found in the category tree: undefined root");
+		}
+		this.insertObjectsOnCache(cache, root, categoryMap);
+	}
+
+	protected void releaseCachedObjects(Cache cache) {
+		List<String> codes = (List<String>) this.get(cache, CATEGORY_CODES_CACHE_NAME, List.class);
+		if (null != codes) {
+			for (int i = 0; i < codes.size(); i++) {

```

Comment:
Is this loop needed? Looks like we could just set CATEGORY_CODES_CACHE_NAME to a new List<String> and the old stuff would be dereferenced. Or is there other stuff in there?

Reply:
It been used to eliminate all the categories, to avoid that they remain categories with dirty references (see the case of the deletion of single category)